File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:
    int solve(int idx, vector<int>& coins, int amount,vector<vector<int>>& dp){
        if(idx == 0){
            if(amount % coins[0] == 0) return amount / coins[0];
            return 1e9; 
        }
        if(dp[idx][amount] != -1) return dp[idx][amount];

        int not_take = solve(idx-1,coins, amount,dp);
        int take = 1e9;
        if(amount >= coins[idx]) take = 1+ solve(idx,coins, amount-coins[idx],dp);
        return dp[idx][amount] = min(take, not_take);
    }

    int coinChange(vector<int>& coins, int amount) {
        int n = coins.size();
        vector<vector<int>> dp(n, vector<int>(amount+1,-1));
        int ans = solve(n-1,coins,amount,dp);
        return ans >= 1e9 ? -1 : ans;

    }
};

File: 0322-coin-change/0322-coin-change.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
var coinChange = function(coins, amount) {
    const n = coins.length
    const dp = Array.from({length: n} , () => Array(amount+1).fill(-1))

    function helper(idx, amount){
        if(idx === 0){
            if(amount % coins[0] === 0) return amount/coins[0]
            return Number.MAX_SAFE_INTEGER;
        }
        if (dp[idx][amount] !== -1) return dp[idx][amount];
        const notTake = helper(idx-1,amount)
        let take = Number.MAX_SAFE_INTEGER
        if(coins[idx] <= amount) take = 1+ helper(idx,amount-coins[idx])

        return dp[idx][amount] = Math.min(take, notTake)
    }

    const ans = helper(n-1, amount)
    return ans >= Number.MAX_SAFE_INTEGER ? -1 : ans
};